### PR TITLE
Fixed card save from slot

### PR DESCRIPTION
**Pull request by VarChar42**
According to https://github.com/RfidResearchGroup/ChameleonUltra/blob/main/docs/protocol.md the command MF1_READ_EMU_BLOCK_DATA accepts 2 arguments: block_start and block_count

### CODE REVIEW

**Comment by VarChar42** [comment]
Also I notived, the function mfClassicGetBlockCount has a parameter isEV1 which is currently not set.

**Comment by VarChar42** [comment]
image
the export from the "Read Card" page seem also broke. There each sector only contains 16 bytes.

**Comment by Foxushka** [comment]
> So only 64 blocks get read from a 1K EV1 card with 72 blocks.

When you load 1k EV1 save into Chameleon, it becomes Mifare Classic 2k, as firmware doesn't support signatures with 1k

**Comment by Foxushka** [comment]
I've pushed fix to write-card branch with commit (cc7e2a2272f07599f89344b24648cf0693b1d7b5). As your fix will lose speed that is critical on mobile. Can you retest it on write-card branch?

**Comment by VarChar42** [comment]
> I've pushed fix to write-card branch with commit ([cc7e2a2](https://github.com/GameTec-live/ChameleonUltraGUI/commit/cc7e2a2272f07599f89344b24648cf0693b1d7b5)). As your fix will lose speed that is critical on mobile. Can you retest it on write-card branch?

Your implementation still has the same problem.
Can you please reopen this PR I will look into speeding it up.

**Comment by Foxushka** [comment]
Are you sure there is issue on last changes in write-card branch?

**Comment by Foxushka** [comment]
Some fixes are in next commit, so only this commit won't work

**Comment by VarChar42** [comment]
> Some fixes are in next commit, so only this commit won't work

yeah including the fixes in the next commit it works.

**Comment by VarChar42** [comment]
my version now checks for block counts not power of 32 .. for example mifare mini with 20 blocks

**Comment by VarChar42** [comment]
Why do you want to copy these changes into your PR? The PR should only contain changes regarding its scope... (my opinion)

**Comment by Foxushka** [comment]
Can you move down count of blocks back to 16, not 32?

**Comment by VarChar42** [comment]
> Can you move down count of blocks back to 16, not 32?

sure. any specific reason for this?

**Comment by Foxushka** [comment]
I don't remember, but I had problems with that, probably unstable via BLE

File: chameleonultragui/lib/gui/menu/slot_export.dart [hunk 1]
```diff
@@ -49,14 +49,44 @@ class SlotExportMenuState extends State<SlotExportMenu> {
...
+      if (remainingBlocks != 0) {
+        Uint8List result = await appState.communicator!
+            .mf1GetEmulatorBlock(blockCount - remainingBlocks, remainingBlocks);
+        binData.setAll(binDataIndex, result);
+      }
+
+      List<Uint8List> sectors = [];
+      int sectorSize = 16 * 4;
+
+      for (int i = 0; i < binData.length; i += sectorSize) {
```
Comment (Foxushka) [inline chameleonultragui/lib/gui/menu/slot_export.dart:83]: I don't actually get this part. Mifare Classic block size is 16 bytes, but you are adding blocks with size 64

Comment (VarChar42) [reply chameleonultragui/lib/gui/menu/slot_export.dart:83]: One sector contains 4 blocks so 64 bytes per sector. Or should one "subarray" only contain one block?

Comment (VarChar42) [reply chameleonultragui/lib/gui/menu/slot_export.dart:83]: so the resulting array is sectors[16][64]

Comment (Foxushka) [reply chameleonultragui/lib/gui/menu/slot_export.dart:83]: resulting array must be only blocks, no sectors. Each block is 16 bytes

Comment (VarChar42) [reply chameleonultragui/lib/gui/menu/slot_export.dart:83]: ok, i will change that

**Review by Foxushka — APPROVED** [review]

**Comment by Foxushka** [comment]
Thank you for fixing this

**Pull request merged**